content/blog/nanoparquet-0-3-0/index.Rmd

```diff
@@ -0,0 +1,341 @@
+
+If the pillar package is loaded, then nanoparquet data frames are
+pretty-printed.
+
+```{r}
+library(nanoparquet)
+library(pillar)
+udf <- system.file("extdata/userdata1.parquet", package = "nanoparquet")
+```
+
+Before actually reading the file, let's look up some metadata about it,
+and also how its columns will be mapped to R types:
+
+```{r}
+parquet_info(udf)
+```
+
+```{r}
+parquet_column_types(udf)
+```
+
+For every Parquet column we see its low level Parquet data type in `type`,
+e.g. `INT64` or `BYTE_ARRAY`. `r_type` the R type that `read_parquet()`
+will create for that column. If `repetition_type` is `REQUIRED`, then that
+column cannot contain missing values. `OPTIONAL` columns may have missing
+values. `logical_type` is the higher level Parquet data type.
+
+E.g. the first column is an UTC (because of the `TRUE`) timestamp, in
+microseconds. It is stored as a 64 bit integer in the file, and it will be
+converted to a `POSIXct` object by `read_parquet()`.
+
+To actually read the file into a data frame, and look at the first
+couple of rows, call
+```{r}
+ud1 <- read_parquet(udf)
+ud1
+```
+
+### Writing a Parquet file
+
+To show `write_parquet()`, we'll use the `flights` data in the nycflights13
+package:
+
+```{r}
+library(nycflights13)
+flights
+```
+
+First we check how columns of `flights` will be mapped to Parquet types:
+
+```{r}
+parquet_column_types(flights)
+```
+
+This looks fine, so we go ahead and write out the file. By default it will
+be Snappy-compressed, and many columns will be dictionary encoded.
+
+```{r}
+write_parquet(flights, "flights.parquet")
+```
+
+### Parquet metadata
+
+Use `parquet_schema()` to see the schema of a Parquet file. The schema
+also includes "internal" parquet columns. Every Parquet file is a tree
+where columns may be part of an "internal" column. nanoparquet currently
+only supports flat files, that consist of a single internal root column
+and all other columns are leaf columns and are children of the root:
+
+```{r}
+parquet_schema("flights.parquet")
+```
+
+To see more information about a Parquet file, use `parquet_metadata()`:
+
+```{r}
+parquet_metadata("flights.parquet")
+```
+
+The output will include the schema, as above, but also data about the
+row groups (`write_parquet()` always writes a single row group currently),
+and column chunks. There is one column chunk per column in each row group.
+
+The columns chunk information also tells you whether a column chunk is
+dictionary encoded, its encoding, its size, etc.
+
+```{r}
+cc <- parquet_metadata("flights.parquet")$column_chunks
+cc[, c("column", "encodings", "dictionary_page_offset")]
+```
+
+```{r}
+cc[["encodings"]][1:3]
+```
+
+## Limitations
+
+nanoparquet 0.3.0 has a number of limitations.
+
+* **Only flat tables**. `read_parquet()` can only read flat tables, i.e.
+  Parquet files where all leaf columns are children of a single root
+  column. Similarly, `write_parquet()` will not write list columns.
```

Comment:
Can you reframe "all leaf columns are children of a single root column" into something more informal that an R user is more likely to understand?